src/typechecker.erl

```diff
@@ -587,6 +587,13 @@ normalize({type, _, union, Tys}, TEnv) ->
         [T] -> T;
         Ts  -> type(union, Ts)
     end;
+normalize({ann_type, _, _} = Type, TEnv) ->
+    Types = flatten_type(Type, TEnv),
+    case merge_union_types(Types, TEnv) of
+        []  -> type(none);
+        [T] -> T;
+        Ts  -> type(union, Ts)
+    end;
```

Comment:
It doesn't work. The absform for an annotated type is `{ann_type, _, List}`, where that third parameter is a list of things. And then `normalize` does nothing on a list, it is expecting a tagged tuple, for all other possibilities (like the list), it passes them to the do-nothing clause at the bottom.

And parsing this list is precisely what `flatten_type` does, so no need to reimplement it 🤷‍♂

Reply:
OK, I had a closer look now. The list is a fixed list of arguments, which is typical for these AST nodes. For `ann_type` it's always 2 arguments, namely the variable name and the type. E.g. the `A :: atom()` (where 55 is a line number) is represented as:

```Erlang
{ann_type,55,[{var,55,'A'},{type,55,atom,[]}]}
```

Using flatten_list and then taking the union of the returned types actually give us the type `A | atom()` instead of the expected `atom()`.

My revised suggestion then:

```Erlang
normalize({ann_type, _Ann, [_Var, Type]}, TEnv) ->
    normalize(Type, TEnv);
```

And I suggest adding a line for this in the existing test case for the normalize function.

Reply:
@zuiderkwast hey there, I'm doing as you suggested, and I haven't seen any infinite loop as described in  #178, but on the other hand I'm having a hard time to follow all the discussion about annotated types like in #180 as well.
What are your thoughts about this?